### PR TITLE
Initial Setup of "Chorus Elasticsearch Edition"

### PR DESCRIPTION
**Pull request by wrigleyDan**
PR for the initial setup of "Chorus Elasticsearch Edition".

The initial setup contains the following components:

* Elasticsearch
* Kibana
* Quepid
* Prometheus
* Grafana

You can find additional information in the README.md file based on the original Chorus project using Solr as the search engine.

### CODE REVIEW

**Comment by epugh** [comment]
I jsut tried with a fresh checkout and did `quickstart.sh` and get `no such service: chorus_es_ui`

**Comment by wrigleyDan** [comment]
Fixed the error, you can give it another try.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      - 3306:3306
+    environment:
+      - MYSQL_ROOT_PASSWORD=password
+    volumes:
+      - ./mysql/docker-entrypoint-initdb.d:/docker-entrypoint-initdb.d
+      - mysql_data:/var/lib/mysql
+
+  elasticsearch:
+    container_name: elasticsearch
+    hostname: elasticsearch
```
Comment (risdenk) [inline docker-compose.yml]: Not needed?

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    ulimits:
+      memlock:
+        soft: -1
+        hard: -1
+    volumes:
+      - ./elasticsearch/elasticsearch.yml:/usr/share/elasticsearch/config/elasticsearch.yml
+      - es_data:/usr/share/elasticsearch/data
+    ports:
+      - "9200:9200"
+      - "9300:9300"
```
Comment (risdenk) [inline docker-compose.yml]: 9300 shouldn't be needed? Its used internally between ES nodes?

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    image: redis:6.2.6-alpine
+    ports:
+      - 6379:6379
+
+  quepid:
+    container_name: quepid
+    image: o19s/quepid:6.10.0
+    ports:
+      - 3000:3000
+    command: "foreman s -f Procfile"
```
Comment (risdenk) [inline docker-compose.yml]: Is the command not specified in the image already?

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      - COOKIES_URL=
+      - EMAIL_MARKETING_MODE=false
+      - EMAIL_PROVIDER=
+      - QUEPID_DEFAULT_SCORER=DCG@10
+      - KEYCLOAK_REALM=chorus
+      - KEYCLOAK_SITE=http://keycloak:9080
+      - SIGNUP_ENABLED=true
+    links:
+      - mysql
+      - redis
```
Comment (risdenk) [inline docker-compose.yml:83]: `depends_on` below will automatically create links

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    container_name: chorus_es_ui
+    build: ./reactivesearch/.
+    volumes:
+      - './reactivesearch:/usr/src/app'
+      - '/usr/src/app/node_modules'
+    ports:
+      - 4001:3000
+
+  redis:
+    container_name: quepid_redis
```
Comment (risdenk) [inline docker-compose.yml]: Don't need this? If you are doing this why not change the whole `redis` section to `quepid_redis`?

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+      - "9300:9300"
+
+  kibana:
+    container_name: kibana
+    image: docker.elastic.co/kibana/kibana:7.16.2
+    ports:
+      - 5601:5601
+
+  chorus-ui:
+    container_name: chorus_es_ui
```
Comment (risdenk) [inline docker-compose.yml]: Don't need this? If you are doing this why not change the whole `chorus-ui` section to `quepid_es_ui`?

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        hard: -1
+    volumes:
+      - ./elasticsearch/elasticsearch.yml:/usr/share/elasticsearch/config/elasticsearch.yml
+      - es_data:/usr/share/elasticsearch/data
+    ports:
+      - "9200:9200"
+      - "9300:9300"
+
+  kibana:
+    container_name: kibana
```
Comment (risdenk) [inline docker-compose.yml:35]: Not needed

File: docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,145 @@
+version: '3'
+services:
+
+  mysql:
+    container_name: mysql
```
Comment (risdenk) [inline docker-compose.yml:5]: Not needed

File: docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    ports:
+      - 9090:9090
+    command:
+      - '--config.file=/etc/prometheus/prometheus.yml'
+    volumes:
+      - ./prometheus/:/etc/prometheus/
+
+  grafana:
+    image: grafana/grafana:7.5.13
+    container_name: grafana
```
Comment (risdenk) [inline docker-compose.yml:98]: Not needed

File: docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    links:
+      - mysql
+      - redis
+    depends_on:
+      - mysql
+      - redis
+
+  prometheus:
+    image: prom/prometheus:v2.32.1
+    container_name: prometheus
```
Comment (risdenk) [inline docker-compose.yml:87]: Not needed

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    environment:
+      - GF_SECURITY_ADMIN_PASSWORD=password
+      - GF_USERS_ALLOW_SIGN_UP=true
+    depends_on:
+      - prometheus
+
+  elasticsearch-exporter:
+    image: quay.io/prometheuscommunity/elasticsearch-exporter:latest
+    hostname: elasticsearch-exporter
+    container_name: elasticsearch-exporter
```
Comment (risdenk) [inline docker-compose.yml]: Not needed

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    command:
+      - '--es.uri=http://elasticsearch:9200'
+    restart: always
+    ports:
+      - "9114:9114"
+
+  keycloak:
+    image: quay.io/keycloak/keycloak:16.1.0
+    container_name: keycloak
+    hostname: keycloak
```
Comment (risdenk) [inline docker-compose.yml]: Not needed?

File: elasticsearch/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,10 @@
+FROM docker.elastic.co/elasticsearch/elasticsearch:7.16.2
+
+WORKDIR /usr/share/elasticsearch
+RUN bin/elasticsearch-plugin install --batch \
+     https://github.com/o19s/elasticsearch-learning-to-rank/releases/download/v1.5.8-es7.16.2/ltr-plugin-v1.5.8-es7.16.2.zip
+RUN ./bin/elasticsearch-plugin install --batch \
+   "https://repo1.maven.org/maven2/org/querqy/querqy-elasticsearch/1.5.es7162.0/querqy-elasticsearch-1.5.es7162.0.zip"
+
+COPY --chown=elasticsearch:elasticsearch elasticsearch.yml /usr/share/elasticsearch/config/
+RUN cat  /usr/share/elasticsearch/config/elasticsearch.yml
```
Comment (risdenk) [inline elasticsearch/Dockerfile]: Why is the `cat` needed?

File: quickstart.sh [hunk 14]
```diff
@@ -0,0 +1,171 @@
+#!/bin/bash
```
Comment (risdenk) [inline quickstart.sh:1]: In general might help to run this through `shellcheck` (same goes for any other `*.sh` script)

**Review by risdenk — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 1 above)
Comment (wrigleyDan) [reply docker-compose.yml]: Removed hostname

File: docker-compose.yml (hunk 2 above)
Comment (wrigleyDan) [reply docker-compose.yml]: Removed port 9300

File: docker-compose.yml (hunk 3 above)
Comment (wrigleyDan) [reply docker-compose.yml]: Indeed, removed the command

File: docker-compose.yml (hunk 4 above)
Comment (wrigleyDan) [reply docker-compose.yml:83]: removed the `links`section

File: docker-compose.yml (hunk 5 above)
Comment (wrigleyDan) [reply docker-compose.yml]: changed `container_name` to `redis`

File: docker-compose.yml (hunk 10 above)
Comment (wrigleyDan) [reply docker-compose.yml:87]: removed hostname

File: docker-compose.yml (hunk 11 above)
Comment (wrigleyDan) [reply docker-compose.yml]: removed hostname

File: docker-compose.yml (hunk 12 above)
Comment (wrigleyDan) [reply docker-compose.yml]: removed hostname

File: elasticsearch/Dockerfile (hunk 13 above)
Comment (wrigleyDan) [reply elasticsearch/Dockerfile]: It is not, copy & paste error, removed the `cat` line

File: quickstart.sh (hunk 14 above)
Comment (wrigleyDan) [reply quickstart.sh:1]: Good suggestions, thanks. No major issues were there, improved a few lines according to the findings. Will run that once we have all components in place.

File: docker-compose.yml (hunk 6 above)
Comment (wrigleyDan) [reply docker-compose.yml]: Changed `container_name` to `chorus_ui`. In general, I like `container_name` definitions to have clear container names and not the ones generated by default, like `chorus-elasticsearch-edition_elasticsearch_1`.

File: docker-compose.yml (hunk 7 above)
Comment (wrigleyDan) [reply docker-compose.yml:35]: Kept `container_name` to have a clear container name and not a generated one.

File: docker-compose.yml (hunk 9 above)
Comment (wrigleyDan) [reply docker-compose.yml:98]: Kept `container_name` to have a clear container name and not a generated one.

File: docker-compose.yml (hunk 8 above)
Comment (wrigleyDan) [reply docker-compose.yml:5]: Kept `container_name` to have a clear container name and not a generated one.

**Pull request merged**